### PR TITLE
Select file on "current" explorer when it exists + file renaming

### PR DESCRIPTION
**Pull request by r4dius**
Added code to select the file on the opened explorer window,
will only open one explorer and reuse it each time,
if file doesn't exist it fallbacks to original open directory thing

### CODE REVIEW

**Comment by r4dius** [comment]
Added file renaming on right click, will do all selected titles, to "standard" :
TITLE_NAME [TITLE_ID][VERSION].extension
- will skip if new filename already exists,
- not sure if it may need more filename sanitizing,
- seems to work other network for me,
- to do : need to update the File Name column and "user.settings" database when renaming happened

**Comment by garoxas** [comment]
Hi thanks for your PR. I have some feedbacks in the review above, most of them are more related to code styling though  

>     * to do : need to update the File Name column and "user.settings" database when renaming happened

I can help with this after you're done with review items above

**Comment by garoxas** [comment]
Oh and it's better if next time you wish to create a PR, you make a new branch for the changes. Because I will squash all commits in PR when merging so your master branch is no longer aligned with the main repo (you can still see 3 commits from your first PR here)

**Comment by r4dius** [comment]
No problem I'll try that next time, I'm not super used to all git operations ^^,
btw I can't find your review feedback am I missing something :/ ?

**Comment by garoxas** [comment]
> btw I can't find your review feedback am I missing something :/ ?

My bad, forgot to click Submit 🤣

File: Windows/Main.Designer.cs [hunk 1]
```diff
@@ -76,6 +76,7 @@ private void InitializeComponent()
             this.olvColumnPermission = ((BrightIdeasSoftware.OLVColumn)(new BrightIdeasSoftware.OLVColumn()));
             this.olvColumnError = ((BrightIdeasSoftware.OLVColumn)(new BrightIdeasSoftware.OLVColumn()));
             this.backgroundWorkerProcess = new System.ComponentModel.BackgroundWorker();
+            this.renameFileToolStripMenuItem = new System.Windows.Forms.ToolStripMenuItem();
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.Designer.cs]: Done :)

File: Windows/Main.Designer.cs [hunk 2]
```diff
@@ -221,7 +222,8 @@ private void InitializeComponent()
             this.contextMenuStrip.Items.AddRange(new System.Windows.Forms.ToolStripItem[] {
             this.copyToolStripMenuItem,
             this.toolStripSeparator4,
-            this.openFileLocationToolStripMenuItem});
+            this.openFileLocationToolStripMenuItem,
+            this.renameFileToolStripMenuItem});
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.Designer.cs]: Done :)

File: Windows/Main.Designer.cs [hunk 3]
```diff
@@ -242,6 +244,13 @@ private void InitializeComponent()
             this.openFileLocationToolStripMenuItem.Size = new System.Drawing.Size(224, 30);
             this.openFileLocationToolStripMenuItem.Text = "Open file locat&ion";
             this.openFileLocationToolStripMenuItem.Click += new System.EventHandler(this.openFileLocationToolStripMenuItem_Click);
+            //
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.Designer.cs]: Done :)

File: Windows/Main.Designer.cs [hunk 4]
```diff
@@ -544,6 +553,7 @@ private void InitializeComponent()
         private BrightIdeasSoftware.OLVColumn olvColumnPermission;
         private BrightIdeasSoftware.OLVColumn olvColumnError;
         private System.ComponentModel.BackgroundWorker backgroundWorkerProcess;
+        private System.Windows.Forms.ToolStripMenuItem renameFileToolStripMenuItem;
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.Designer.cs]: Done :)

File: Windows/Main.cs [hunk 5]
```diff
@@ -845,18 +845,37 @@ private void copyToolStripMenuItem_Click(object sender, EventArgs e)
             contextMenuStrip.Tag = null;
         }
 
+        [DllImport("shell32.dll", CharSet = CharSet.Unicode)]
+        private static extern IntPtr ILCreateFromPathW(string pszPath);
+
+        [DllImport("shell32.dll")]
+        private static extern int SHOpenFolderAndSelectItems(IntPtr pidlFolder, int cild, IntPtr apidl, int dwFlags);
+
+        [DllImport("shell32.dll")]
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.cs]: Doooone :)

File: Windows/Main.cs [hunk 6]
```diff
@@ -867,6 +886,32 @@ private void openFileLocationToolStripMenuItem_Click(object sender, EventArgs e)
             contextMenuStrip.Tag = null;
         }
 
+        private void renameFileToolStripMenuItem_Click(object sender, EventArgs e)
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.cs]: Added a bunch of conditions and counting how many files will be renamed / are missing / already exists / don't need renaming,
First loop now counts everything, if 0 file can be renamed we messagebox the counting results,
if 1 or more files can be renamed we ask for confirmation first and show the counting after renaming.

Would need some testing, I may have missed some conditions ^^

[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.cs]: May need some new form as messagebox dialog for a better looking popup, I will test it eventually

File: Windows/Main.cs [hunk 7]
```diff
@@ -867,6 +886,32 @@ private void openFileLocationToolStripMenuItem_Click(object sender, EventArgs e)
...
+        {
+            if (contextMenuStrip.Tag is Title title)
+            {
+                foreach(OLVListItem item in objectListView.SelectedItems)
```
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.cs]: Done :)

File: Windows/Main.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (r4dius) [reply Windows/Main.cs]: "Optimized" some messages but still not sure of the phrasing, fill free to edit, I'll be looking at the windows messages style

[the remark this answers is not in the document]
Comment (garoxas) [reply Windows/Main.cs]: > 
> 
> "Optimized" some messages but still not sure of the phrasing, fill free to edit, I'll be looking at the windows messages style

I usually decide on wording from similar functions in other popular apps, and if Windows has it then for sure I'll use theirs word by word
I've already replaced some strings but still looking for better alternatives for the rest of them

[the remark this answers is not in the document]
Comment (garoxas) [reply Windows/Main.cs]: Submitted my changes. The changes look massive but most of them are simply renaming variables and strings

**Comment by garoxas** [comment]
Everything looks okay. Aside from updating file name column and in the user.settings but that can be done later
If you have nothing more to add I can merge this now 😃

**Comment by r4dius** [comment]
> Everything looks okay. Aside from updating file name column and in the user.settings but that can be done later
> If you have nothing more to add I can merge this now 😃

Looks nicer for sure ^^ no problem for me, will test after the merge :)

**Comment by garoxas** [comment]
Thank you for the PR. Much appreciated

**Comment by r4dius** [comment]
;)

**Pull request merged**